addons/binding/org.openhab.binding.roomba/src/main/java/org/openhab/binding/roomba/discovery/RoombaDiscoveryService.java

```diff
@@ -0,0 +1,132 @@
+import org.openhab.binding.roomba.model.discovery.RoombaDiscoveryResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+/**
+ * Discovers Roombas that are available on the network. This works for models 980+.
+ *
+ * Roombas are discovered by sending "irobotmcs" to 255.255.255.255 via UDP Broadcast. The Roomba must already be paired
+ * to the WiFi network.
+ *
+ * @author Stephen Liang
+ */
+public class RoombaDiscoveryService extends AbstractDiscoveryService {
+    private final static Logger logger = LoggerFactory.getLogger(RoombaDiscoveryService.class);
+
+    public RoombaDiscoveryService() throws IllegalArgumentException {
+        super(SUPPORTED_THING_TYPES_UIDS, 15, false);
+    }
+
+    /**
+     * Scans for roombas by spawning a search thread
+     */
+    @Override
+    protected void startScan() {
+        logger.info("Starting discovery for iRobot Roombas");
+        scheduler.execute(searchRunnable);
+    }
+
+    // Runnable for search adapter
+
+    private Runnable searchRunnable = new Runnable() {
+
+        @Override
+        public void run() {
+            logger.debug("Start adapter discovery ");
+            logger.trace("Send broadcast message");
+            DatagramSocket localSocket = null;
+
+            try {
+                localSocket = new DatagramSocket();
+                localSocket.setBroadcast(true);
+                localSocket.setSoTimeout(10000); // Listen 10 seconds
+
+                byte[] sendData = BROADCAST_MESSAGE.getBytes();
+
+                DatagramPacket sendPacket = new DatagramPacket(sendData, sendData.length,
+                        InetAddress.getByName(BROADCAST_ADDRESS), BROADCAST_PORT);
+                localSocket.send(sendPacket);
+
+                byte[] receiveBuffer = new byte[1000];
+
+                // Listen for answer
+
+                DatagramPacket receivePacket = new DatagramPacket(receiveBuffer, receiveBuffer.length);
+                localSocket.receive(receivePacket);
+                String receiveMessage = new String(receivePacket.getData()).trim();
+                String receiveIP = receivePacket.getAddress().getHostAddress();
+                int receivePort = receivePacket.getPort();
+                logger.trace("Received Message: {}  ", receiveMessage);
+                logger.trace("Received from Host: {}", receiveIP);
+                logger.trace("Received from Port: {}", receivePort);
+
+                // Parse the response (json) into its POJO equivalent
+                Gson gson = new Gson();
+                RoombaDiscoveryResult discoveryResult = gson.fromJson(receiveMessage, RoombaDiscoveryResult.class);
+
+                // Add the roomba as a discovery result
+                addRoomba(discoveryResult);
+            } catch (IOException e) {
+                logger.info("Unable to find any Roomba!", e);
+            } finally {
+                try {
```

Comment:
This is proper way of closing 👍 

You could consider using a library for this I believe its already used somewhere https://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/IOUtils.html#closeQuietly(java.io.Closeable...)